packages/service-library/src/servicelib/fastapi/requests_decorators.py

```diff
@@ -1,110 +1,117 @@
-
-        @cancellable_request
-        async def route(
-            _request: Request,
-            ...
-        )
+#
+# cancel_on_disconnect/disconnect_poller based
+# on https://github.com/RedRoserade/fastapi-disconnect-example/blob/main/app.py
+#
+_POLL_INTERVAL_S: float = 0.01
+
+
+async def disconnect_poller(request: Request, result: Any):
+    """
+    Poll for a disconnect.
+    If the request disconnects, stop polling and return.
     """
-    # CHECK: Early check that will raise upon import
-    # IMPROVEMENT: inject this parameter to handler_fun here before it returned in the wrapper and consumed by fastapi.router?
-    found_required_arg = any(
-        parameter.name == "_request" and parameter.annotation == Request
-        for parameter in inspect.signature(handler_fun).parameters.values()
+    while not await request.is_disconnected():
+        await asyncio.sleep(_POLL_INTERVAL_S)
+
+    logger.debug(
+        "client %s disconnected! Cancelling handler for request %s %s",
+        request.client,
+        request.method,
+        request.url,
     )
-    if not found_required_arg:
-        raise ValueError(
-            f"Invalid handler {handler_fun.__name__} signature: missing required parameter _request: Request"
-        )
+    return result
+
+
+def cancel_on_disconnect(handler: _HandlerWithRequestArg):
 
-    # WRAPPER ----
-    @wraps(handler_fun)
-    async def wrapper(*args, **kwargs) -> Optional[Any]:
-        request: Request = kwargs["_request"]
+    _validate_signature(handler)
 
-        # Intercepts handler call and creates a task out of it
+    @wraps(handler)
+    async def wrapper(request: Request, *args, **kwargs):
+        sentinel = object()
+
+        # Create two tasks:
+        # one to poll the request and check if the client disconnected
+        poller_task = asyncio.create_task(
+            disconnect_poller(request, sentinel),
+            name=f"cancel_on_disconnect/poller/{handler.__name__}/{id(sentinel)}",
+        )
+        # , and another which is the request handler
         handler_task = asyncio.create_task(
-            handler_fun(*args, **kwargs),
-            name=f"{TASK_NAME_PREFIX}/handler/{handler_fun.__name__}",
+            handler(request, *args, **kwargs),
+            name=f"cancel_on_disconnect/handler/{handler.__name__}/{id(sentinel)}",
         )
-        # An extra task to monitor when the client disconnects so it can
-        # cancel 'handler_task'
-        auto_cancel_task = asyncio.create_task(
-            _cancel_task_if_client_disconnected(request, handler_task),
-            name=f"{TASK_NAME_PREFIX}/auto_cancel/{handler_fun.__name__}",
+
+        done, pending = await asyncio.wait(
+            [poller_task, handler_task], return_when=asyncio.FIRST_COMPLETED
         )
 
-        try:
+        # One has completed, cancel the other
+        for t in pending:
+            t.cancel()
+            try:
+                await t
```

Comment:
One thing to consider (might not be the case here). If you have something that can end up in a deadlock (uses network or other usually locked resources). Add a timeout when cancelling.

I had that issue with Redis which caused a lot of headaches. See below for more details

https://github.com/ITISFoundation/osparc-simcore/blob/373c3584b90d187ce7468d6e68690abcfc7a9182/services/director-v2/src/simcore_service_director_v2/modules/redis.py#L137-L156

Reply:
Although in this case I am be puzzled and wonder whether timing out will actually execute the cancelation ...